### PR TITLE
Fixed #27003: ArrayField and JSONField form fields check for already-converted values

### PR DESCRIPTION
**Pull request by brandonchinn178**
`ArrayField`'s and `JSONField`'s `to_python` method goes straight to parsing the value, instead of checking if the value has already been converted to its type.

### CODE REVIEW

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: tests/postgres_tests/test_json.py [hunk 1]
```diff
@@ -309,3 +309,18 @@ class CustomJSONField(forms.JSONField):
...
+            ['a', 'b', 'c'],
+            {'a': 1, 'b': 2},
+            1,
+            True,
+            None,
+        ]
+        for val in vals:
+            try:
+                self.assertEqual(field.clean(val), val)
+            except ValueError:
```
Comment (charettes) [inline tests/postgres_tests/test_json.py]: Shouldn't it be a `ValidationError` in this case? What about letting the original exception propagate instead?

File: tests/postgres_tests/test_array.py [hunk 2]
```diff
@@ -633,6 +633,14 @@ def test_model_field_formfield_size(self):
         self.assertIsInstance(form_field, SimpleArrayField)
         self.assertEqual(form_field.max_length, 4)
 
+    def test_already_converted_value(self):
+        field = SimpleArrayField(forms.CharField())
+        vals = ['a', 'b', 'c']
+        try:
+            self.assertEqual(field.clean(vals), vals)
+        except AttributeError:
```
Comment (charettes) [inline tests/postgres_tests/test_array.py]: Ditto.

File: tests/postgres_tests/test_json.py (hunk 1 above)
Comment (timgraham) [inline tests/postgres_tests/test_json.py]: Yes, we don't need a try/except/fail. Hiding the original exception makes debugging more difficult and I've removed this pattern wherever I found it in Django's test suite.

Please see the build failure for the JSON test failure and uncheck "Patch needs improvement" on the ticket when things are passing.

**Comment by In-vincible** [comment]
I am a total beginner in django development, I have gone through few of the tutorials like Writing your first patch for Django,Submitting patches, I want to participate/help in this bug fix, right now I am trying to grasp the fix given by brandonchinn178, this bug is not assigned to anyone right now, how can I assign it to myself?

**Comment by timgraham** [comment]
There's nothing to do here except to review the patch. You don't need to assign the ticket to do that, but claiming tickets is described at https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#claiming-tickets.

File: django/contrib/postgres/forms/jsonb.py [hunk 3]
```diff
@@ -11,6 +11,10 @@ class InvalidJSONInput(six.text_type):
     pass
 
 
+class JsonStr(six.text_type):
```
Comment (frgtn) [inline django/contrib/postgres/forms/jsonb.py]: _Minor:_ `JsonStr` could be renamed to `JSONStr` or `JSONString` to be consistent with the naming of `InvalidJSONInput` class above.

File: django/contrib/postgres/forms/jsonb.py [hunk 4]
```diff
@@ -22,15 +26,23 @@ def to_python(self, value):
...
+            converted = json.loads(value)
         except ValueError:
             raise forms.ValidationError(
                 self.error_messages['invalid'],
                 code='invalid',
                 params={'value': value},
             )
 
+        if isinstance(converted, six.text_type):
+            return JsonStr(converted)
```
Comment (frgtn) [inline django/contrib/postgres/forms/jsonb.py]: @mjtamlyn can you take a look if returning a subclass of `six.text_type` could potentially break things somewhere down the line?

File: AUTHORS [hunk 5]
```diff
@@ -110,6 +110,7 @@ answer newbie questions, and generally made Django that much better:
     Božidar Benko <bbenko@gmail.com>
     Brad Melin <melinbrad@gmail.com>
     Brant Harris
+    Brandon Chinn <http://brandonchinn178.github.io>
```
Comment (timgraham) [inline AUTHORS]: Brandon sorts after Brendan.

Comment (brandonchinn178) [inline AUTHORS]: lol duh

**Comment by brandonchinn178** [comment]
@timgraham what do you mean "split it into two commits"? What are you referring to?

**Comment by timgraham** [comment]
There are currently 11 commits on this branch. It would be better to merge this as two commits, one commit for the fix for `ArrayField` and one for fix for `JSONField`.

**Comment by brandonchinn178** [comment]
So squash commits? Or do you mean two pull requests?

**Comment by timgraham** [comment]
Squash

**Comment by timgraham** [comment]
merged in 6573274161ee589e89077192239e0bd01cbd692a and eed615000903ec7b659c3563b7b4b4b3f34f5636, thanks!

**Pull request closed without merging**